core/src/main/java/org/infinispan/statetransfer/StateProviderImpl.java

```diff
@@ -207,7 +211,11 @@ private CacheTopology getCacheTopology(int requestTopologyId, Address destinatio
                   "topology (%d). Waiting for topology %d to be installed locally.", isReqForTransactions ? "Transactions" : "Segments", destination,
                   requestTopologyId, currentTopologyId, requestTopologyId);
          }
-         stateTransferLock.waitForTopology(requestTopologyId, timeout, TimeUnit.MILLISECONDS);
+         try {
+            stateTransferLock.waitForTopology(requestTopologyId, timeout, TimeUnit.MILLISECONDS);
+         } catch (TimeoutException e) {
+            throw new CacheException("Failed waiting for topology " + requestTopologyId, e);
```

Comment:
Please add a log message to return an `o.i.u.c.TimeouException`, and use that everywhere `waitForTopology`/`topologyFuture` is called.